src/main/java/seedu/address/model/animal/UniqueAnimalList.java

```diff
@@ -0,0 +1,150 @@
+        }
+        internalList.add(toAdd);
+    }
+
+    /**
+     * Replaces the animal {@code target} in the list with {@code editedAnimal}.
+     * {@code target} must exist in the list.
+     * The animal identity of {@code editedAnimal} must not be the same as another existing animal in the list.
+     */
+    public void setAnimal(Animal target, Animal editedAnimal) {
+        requireAllNonNull(target, editedAnimal);
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            throw new AnimalNotFoundException();
+        }
+
+        if (!target.isSameAnimal(editedAnimal) && contains(editedAnimal)) {
+            throw new DuplicateAnimalException();
+        }
+
+        internalList.set(index, editedAnimal);
+    }
+
+    /**
+     * Removes the equivalent animal from the list.
+     * The animal must exist in the list.
+     */
+    public void remove(Animal toRemove) {
+        requireNonNull(toRemove);
+        if (!internalList.remove(toRemove)) {
+            throw new AnimalNotFoundException();
+        }
+    }
+
+    public void setAnimals(UniqueAnimalList replacement) {
+        requireNonNull(replacement);
+        internalList.setAll(replacement.internalList);
+    }
+
+    /**
+     * Replaces the contents of this list with {@code animals}.
+     * {@code animals} must not contain duplicate animals.
+     */
+    public void setAnimals(List<Animal> animals) {
+        requireAllNonNull(animals);
+        if (!animalsAreUnique(animals)) {
+            throw new DuplicateAnimalException();
+        }
+
+        internalList.setAll(animals);
+    }
+
+    /**
+     * Returns the backing list as an unmodifiable {@code ObservableList}.
+     */
+    public ObservableList<Animal> asUnmodifiableObservableList() {
+        return internalUnmodifiableList;
+    }
+
+    @Override
+    public Iterator<Animal> iterator() {
+        return internalList.iterator();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof UniqueAnimalList)) {
+            return false;
+        }
+
+        UniqueAnimalList otherUniquePersonList = (UniqueAnimalList) other;
+        return internalList.equals(otherUniquePersonList.internalList);
+    }
+
+    @Override
+    public int hashCode() {
+        return internalList.hashCode();
+    }
+
+    @Override
+    public String toString() {
+        return internalList.toString();
+    }
+
+    /**
+     * Returns true if {@code animals} contains only unique animals.
+     */
+    private boolean animalsAreUnique(List<Animal> animals) {
+        for (int i = 0; i < animals.size() - 1; i++) {
+            for (int j = i + 1; j < animals.size(); j++) {
+                if (animals.get(i).isSameAnimal(animals.get(j))) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
```

Comment:
This method could be simplified to:
```java
// list of distinct animals and original list are same size => only unique animals
animals.stream().distinct().count() == animals.size()
```

Reply:
But this suggested change is that it uses `Animal::equals` instead of `Animal::isSameAnimal`. Not sure if that would result in an unwanted side-effect. Perhaps this change is not needed.